hello-compose-app/src/main/kotlin/hello/com/pose/MainScreen.kt

```diff
@@ -15,42 +13,82 @@ import androidx.compose.ui.unit.dp
 import androidx.lifecycle.viewmodel.compose.viewModel
 import coil.ImageLoader
 import coil.compose.rememberAsyncImagePainter
+import hello.com.pose.composable.Alert
 import hello.com.pose.composable.SearchBar
 import hello.com.pose.shared.domain.photo.Photo
+import kotlinx.coroutines.launch
 
 @Composable
 fun MainScreen(viewModel: MainViewModel = viewModel()) {
-    val query: MutableState<String> = remember { mutableStateOf("") }
+    val lazyGridState = rememberLazyGridState()
+    val coroutineScope = rememberCoroutineScope()
+    val currentSearchQuery by viewModel.searchQuery.collectAsState()
+
+    val shouldStartPaginate = remember {
+        derivedStateOf {
+            viewModel.canPaginate && lazyGridState.isLastItemVisible
+        }
+    }
+
+    LaunchedEffect(key1 = shouldStartPaginate.value) {
+        if (shouldStartPaginate.value && viewModel.pagingState == PagingState.IDLE) {
+            viewModel.getPhotosByQuery(currentSearchQuery)
+        }
+    }
 
     Surface(modifier = Modifier.fillMaxSize()) {
         Column(modifier = Modifier.padding(8.dp)) {
-            SearchBar(text = query.value,
+            SearchBar(text = currentSearchQuery,
                 inputChange = {
-                    query.value = it
-                    viewModel.getImageList(query.value)
+                    viewModel.setNewQuery(it)
                 })
-
-            LazyVerticalGrid(columns = GridCells.Fixed(4)) {
-                viewModel.photos.value.data.let { photo ->
-                    items(photo) {
-                        MainContentItem(photo = it)
+            SearchList(viewModel, lazyGridState)
+            LaunchedEffect(key1 = currentSearchQuery) {
+                if (viewModel.prevQuery != currentSearchQuery) {
+                    coroutineScope.launch {
+                        lazyGridState.animateScrollToItem(index = 0)
                     }
                 }
             }
         }
     }
 }
 
+@Composable
+fun SearchList(viewModel: MainViewModel, gridState: LazyGridState) {
```

Comment:
ViewModel보다 리스트를 인자로 변경해서 상태를 좀더 끌어올리는게 좋아보입니다!

Reply:
좋습니다! 리스트를 인자로 가지도록 수정하겠습니다
```suggestion
fun SearchList(photos: MutableList<Photo>, gridState: LazyGridState) {
```